flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecDeduplicate.scala

```diff
@@ -73,4 +93,93 @@ class StreamExecDeduplicate(
       .item("order", orderString)
   }
 
+  //~ ExecNode methods -----------------------------------------------------------
+
+  override protected def translateToPlanInternal(
+      tableEnv: StreamTableEnvironment): StreamTransformation[BaseRow] = {
+
+    // FIXME checkInput is not acc retract after FLINK-12098 is done
+    val inputIsAccRetract = false
+
+    if (inputIsAccRetract) {
+      throw new TableException(
+        "Deduplicate: Retraction on Deduplicate is not supported yet.\n" +
+          "please re-check sql grammar. \n" +
+          "Note: Deduplicate should not follow a non-windowed GroupBy aggregation.")
+    }
+
+    val inputTransform = getInputNodes.get(0).translateToPlan(tableEnv)
+      .asInstanceOf[StreamTransformation[BaseRow]]
+
+    val rowTypeInfo = inputTransform.getOutputType.asInstanceOf[BaseRowTypeInfo]
+
+    // FIXME infer generate retraction after FLINK-12098 is done
+    val generateRetraction = true
```

Comment:
val generateRetraction = producesUpdates?
Add some it case to test no retraction in first row?

Reply:
whether generate Retraction could be inferred based on Retraction rules, which is done in FLINK-12098. I could add a TODO message here.